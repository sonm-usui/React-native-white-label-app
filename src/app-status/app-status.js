// @flow
import React from 'react'
import { connect } from 'react-redux'
import { AppState, StyleSheet, View } from 'react-native'
import { BlurView } from '@react-native-community/blur'

import { getUnacknowledgedMessages } from './../store/config-store'
import { getRestoreStatus } from './../store/store-selector'
import type { Store } from './../store/type-store'
import { RestoreStatus } from '../restore/type-restore'
import type {
  AppStatusProps,
  AppStatusState,
  ConnectProps,
} from './type-app-status'
import {
  resetBackgroundTimeout,
  watchApplicationInactivity,
} from '../bridge/react-native-cxs/RNCxs'

export class AppStatusComponent extends React.Component<
  AppStatusProps,
  AppStatusState
> {
  state = {
    appState: AppState.currentState,
  }

  subscription = null;

  componentDidMount() {
    this.subscription = AppState.addEventListener('change', this._handleAppStateChange)
  }

  componentWillUnmount() {
    this.subscription.remove()
  }

  _handleAppStateChange = (nextAppState) => {
    if (
      this.state.appState &&
      this.state.appState.match(/inactive|background/) &&
      nextAppState === 'active' &&
      (this.props.restoreStatus === RestoreStatus.none ||
        this.props.restoreStatus === RestoreStatus.RESTORE_DATA_STORE_SUCCESS)
    ) {
      this.props.dispatch(getUnacknowledgedMessages())
    }
    this.setState({ appState: nextAppState })
  }

  componentDidUpdate() {
    if (
      this.state.appState &&
      this.state.appState.match(/inactive|background/)
    ) {
      watchApplicationInactivity().then((r) => r)
    } else {
      resetBackgroundTimeout()
    }
  }

  render() {
    if (
      this.state.appState &&
      this.state.appState.match(/inactive|background/)
    ) {
      return (
        <View style={styles.container}>
          <BlurView
            style={styles.blurView}
            // $FlowFixMe
            reducedTransparencyFallbackColor={String('gray')}
            blurType="light"
            blurAmount={20}
          />
        </View>
      )
    }

    return null
  }
}

const styles = StyleSheet.create({
  container: {
    zIndex: 1,
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
  },
  blurView: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
  },
})

const mapStateToProps = (state: Store): ConnectProps => {
  return {
    restoreStatus: getRestoreStatus(state),
  }
}

const AppStatus = connect<AppStatusProps, {||}, _, _, _, _>(mapStateToProps)(
  AppStatusComponent
)

export default AppStatus
