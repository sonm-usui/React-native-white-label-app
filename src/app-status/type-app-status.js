// @flow

import type { ReduxConnect } from '../common/type-common'

export type AppStatusState = {
  appState: ?string,
}

export type AppStatusProps = {
  ...ConnectProps,
} & ReduxConnect

export type ConnectProps = {|
  restoreStatus: string,
|}
