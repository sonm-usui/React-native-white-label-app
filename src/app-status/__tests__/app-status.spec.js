// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'

import { AppStatusComponent } from '../app-status'

const getProps = () => {
  return {
    restoreStatus: 'restoreStatus',
    dispatch: jest.fn(),
  }
}

const setup = () => {
  const props = getProps()
  const component = renderer.create(<AppStatusComponent {...props} />)
  const instance = component.root.instance

  return { component, instance }
}

describe('<AppStatusComponent />', () => {
  it('AppStatusComponent with blur snapshot', () => {
    const { component, instance } = setup()
    instance.setState({ appState: 'background' })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('AppStatusComponent without blur snapshot', () => {
    const { component, instance } = setup()
    instance.setState({ appState: 'active' })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
