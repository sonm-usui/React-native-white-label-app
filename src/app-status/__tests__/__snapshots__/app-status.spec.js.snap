// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppStatusComponent /> AppStatusComponent with blur snapshot 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
/>
`;

exports[`<AppStatusComponent /> AppStatusComponent without blur snapshot 1`] = `null`;
